snuba/state/rate_limit.py

```diff
@@ -27,6 +31,57 @@
 rds = get_redis_client(RedisClientKey.RATE_LIMITER)
 
 
+def get_rate_limit_config(
+    per_second: tuple[str, float | None], concurrent: tuple[str, int | None]
+) -> tuple[Any, Any]:
+    """
+    This function to encapsulate how rate limit keys are fetched from Redis, since
+    that is conceptually a different process from fetching normal config keys.
+
+    It currently also contains logic to read from both the new namespace and the old one,
+    until everything is properly migrated.
+
+    First step: read from the old and new namespace. If the new namespace is different from
+    the old, copy the value from old to new. Return the values from the old namespace.
+    """
+    ps_name, per_second_default = per_second
+    ct_name, concurrent_default = concurrent
+
+    # Dual read
+    # if the value in new doesn't match value from old, copy over
+    old_per_second, old_concurrent = get_configs([(ps_name, None), (ct_name, None)])
+    new_per_second, new_concurrent = get_configs(
+        [(ps_name, None), (ct_name, None)], config_key=state.rate_limit_config_hash
+    )
+
+    # This handles deletes as well, since writing None deletes the key
+    if old_per_second != new_per_second:
+        set_rate_limit_config(ps_name, old_per_second)
+    if old_concurrent != new_concurrent:
+        set_rate_limit_config(ct_name, old_concurrent)
+
+    found_per_second = (
+        old_per_second if old_per_second is not None else per_second_default
+    )
+    found_concurrent = (
+        old_concurrent if old_concurrent is not None else concurrent_default
+    )
+
+    return (found_per_second, found_concurrent)
+
+
+def set_rate_limit_config(bucket: str, value: float | int | None) -> None:
+    """
+    This function to encapsulate how rate limit keys are set in Redis, since
+    that is conceptually a different process from fetching normal config keys.
+    """
+    set_config(
+        bucket,
+        value,
+        config_key=state.rate_limit_config_hash,
```

Comment:
if you're setting with `config_key=state.rate_limit_config_hash` but your `get_rate_limit_config` defaults to old values, are you not going to override whatever you set after you deploy this PR? Maybe it would make sense to write to both?

Reply:
I'm not sure I follow which case you're referring to here?

Reply:
Here's a test that fails which I think should pass:

```python3
def test_overwrite_new_with_old():
    ps_key = "project_per_second_limit_36"
    ct_key = "project_concurrent_limit_36"

    set_rate_limit_config(ps_key, 123)
    set_rate_limit_config(ct_key, 321)

    ps_found, ct_found = get_rate_limit_config((ps_key, 76), (ct_key, 77))
    assert ps_found == 123
    assert ct_found == 321
```


After re-reading this PR, I understand what is happening but it is pretty confusing. As I read this module, I see two public functions:

```
def get_rate_limit_config
def set_rate_limit_config
```

As a reader, I assume that rate limits are retrieved by `get_rate_limit_config` and set by `set_rate_limit_config`.

That is not actually true with the code in this PR. `set_rate_limit_config` is an internal function called in this module which sets the configuration with a namespace, however the way rate limits are _actually_ set is by calling `state.set_config` without a `config_key`, thus putting it by default into the old default namespace. I understand this is maybe an interim state and maybe we could document it, but I think it would just be easier if this module's getters and setters worked symmetrically.

Reply:
Oh that's a good point. I'll change that.